src/app/binding/BindingTable.cpp

```diff
@@ -0,0 +1,237 @@
+    // Last binding with the peer, close the connection.
+    if (peerNumBindings == 1)
+    {
+        OperationalDeviceProxy * device = FindPeerDevice(peer);
+        if (device)
+        {
+            mPeerDevices.ReleaseObject(device);
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingTable::TryConnectDevice(OperationalDeviceProxy * device)
+{
+    if (device == nullptr)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    if (device->IsConnected())
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    if (device->IsConnecting())
+    {
+        return CHIP_NO_ERROR;
+    }
+    Dnssd::Resolver::Instance().SetResolverDelegate(&mResolveDelegate);
+    CHIP_ERROR ret = Dnssd::Resolver::Instance().ResolveNodeId(device->GetPeerId(), Inet::IPAddressType::kAny);
+    return ret;
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(NodeId node, FabricIndex fabricIndex)
+{
+    return FindBindingEntry(BuildPeerId(node, fabricIndex));
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(PeerId peer)
+{
+    BindingTableEntry * foundEntry = nullptr;
+    mBindingEntries.ForEachActiveObject([&foundEntry, peer](BindingTableEntry * entry) {
+        if (entry->mPeerId == peer)
+        {
+            foundEntry = entry;
+            return false;
+        }
+        return true;
+    });
+    return foundEntry;
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(EndpointId endpoint, ClusterId cluster)
+{
+    BindingTableEntry * foundEntry = nullptr;
+    mBindingEntries.ForEachActiveObject([&foundEntry, endpoint, cluster](BindingTableEntry * entry) {
+        if (entry->mSourceEndpoint == endpoint && entry->mClusterId == cluster)
+        {
+            foundEntry = entry;
+            return false;
+        }
+        return true;
+    });
+    return foundEntry;
+}
+
+OperationalDeviceProxy * BindingTable::FindPeerDevice(NodeId node, FabricIndex fabricIndex)
+{
+    return FindPeerDevice(BuildPeerId(node, fabricIndex));
+}
+
+OperationalDeviceProxy * BindingTable::FindPeerDevice(PeerId peer)
+{
+    OperationalDeviceProxy * foundDevice = nullptr;
+    mPeerDevices.ForEachActiveObject([&foundDevice, peer](OperationalDeviceProxy * device) {
+        if (device->GetPeerId() == peer)
+        {
+            foundDevice = device;
+            return false;
+        }
+        return true;
+    });
+    return foundDevice;
+}
+
+void BindingTable::ResolveDelegate::OnNodeIdResolved(const Dnssd::ResolvedNodeData & nodeData)
+{
+    OperationalDeviceProxy * device = mBindingTable->FindPeerDevice(nodeData.mPeerId);
+    if (nodeData.mNumIPs > 0 && device != nullptr && !device->IsConnected() && !device->IsConnecting())
+    {
+        ChipLogProgress(AppServer, "BindingTable: NodeId resolved for 0xdevice %" PRIX64, nodeData.mPeerId.GetNodeId());
+        Transport::PeerAddress addr(nodeData.mAddress[0], Transport::Type::kUdp);
+        addr.SetInterface(nodeData.mInterfaceId);
```

Comment:
This is wrong.  We should not be using the interface id, except for link-local ipv6 addresses.

See what `DeviceController::ToPeerAddress` does.  We should move that into a utility of some sort....